apex/contrib/csrc/group_norm/group_norm_nhwc_op.cpp

```diff
@@ -0,0 +1,243 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ *modification, are not permit- ted.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY DIRECT,
+ *INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ *DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ *OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ *NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+ *EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ **************************************************************************************************/
+
+#include <ATen/cuda/CUDAContext.h>
+#include <torch/extension.h>
+
+#include "group_norm_nhwc_fwd_one_pass.h"
+#include "group_norm_nhwc_bwd_one_pass.h"
+#include "group_norm_nhwc.h"
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#define CHECK_CUDA_STATUS(call)                                       \
+  do {                                                                \
+    cudaError_t status_ = call;                                       \
+    if (status_ != cudaSuccess) {                                     \
+      fprintf(stderr, "CUDA error (%s:%d): %s\n", __FILE__, __LINE__, \
+              cudaGetErrorString(status_));                           \
+      exit(1);                                                        \
+    }                                                                 \
+  } while (0)
+
+#define CHECK_CUDA(x) \
+  AT_ASSERTM(x.type().is_cuda(), #x " must be a CUDA tensor")
+#define CHECK_CONTIGUOUS(x) \
+  AT_ASSERTM(x.is_contiguous(), #x " must be contiguous")
+#define CHECK_CHANNELS_LAST(x)                                \
+  AT_ASSERTM(x.is_contiguous(at::MemoryFormat::ChannelsLast), \
+             #x " must be channels last")
+#define CHECK_INPUT(x) \
+  CHECK_CUDA(x);       \
+  CHECK_CONTIGUOUS(x)
+#define CHECK_NHWC_INPUT(x) \
+  CHECK_CUDA(x);            \
+  CHECK_CHANNELS_LAST(x)
+
+enum Algo { OnePass, TwoPass };
+
+static bool initialized = false;
+static cudaDeviceProp props;
+
+std::vector<torch::Tensor> group_norm_fwd(torch::Tensor input, int groups,
+                                          torch::Tensor weight,
+                                          torch::Tensor bias, float eps,
+                                          Algo algo, 
+                                          bool with_swish = false) {
+  if (!initialized) {
+    CHECK_CUDA_STATUS(cudaGetDeviceProperties(&props, 0));
+    initialized = true;
+  }
+  CHECK_NHWC_INPUT(input);
```

Comment:
I prefer to explicitly throw an error if we're sure about the supported values of `groups`

```c++
  TORCH_CHECK(supported_groups_values.count(groups), "`groups` of 32 is only supported but ", groups, " is passed");
  CHECK_NHWC_INPUT(input);
```

Reply:
Thanks, all the checking has been added to the Python side and unsupported cases will be forwarded to PyTorch implementation.

Reply:
@alpha0422 That's great but still I think it'd be better to have a check in C++ side as well